manuals/nk2info.1
.Dd June 25, 2026
.Dt NK2INFO 1
.Os
.Sh NAME
.Nm nk2info
.Nd determines information about a Nickfile (NK2)
.Sh SYNOPSIS
.Nm nk2info
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm nk2info
is a utility to determine information about a Nickfile (NK2)
.Pp
.Nm nk2info
is part of the
.Nm libnk2
package.
.Nm libnk2
is a library to access the Nickfile (NK2) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# nk2info Outlook.NK2
nk2info 20260625
.sp
Nickfile information:
	Last modification time	: Jul 04, 2011 09:56:01.557000 UTC
	Number of aliases	: 4
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr nk2export 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libnk2/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* nk2export check if always adding .export is the desired behavior
* libnk2_io_handle.c check if LIBNK2_VALUE_TYPE_BOOLEAN is working as intended
* item values
  - implement currency as: LIBFVALUE_VALUE_TYPE_FIXED_POINT_64BIT
    what about location of the point ?
* in io_handle_read_items used set array value instead of append ?

Optimizations:
* map (file offset and size) items and item values first read on demand ?
* pre allocate item array entries

library
* additional value type support functions
  - filetime
* what about multiple alias entries 0x6001 of a single item?
  - they are set in values table but identifier lookup will return first only
* local library support

Tools
* add signal handling
* nk2export
  - add export mode like pffexport
  - export alias files instead of item values
  - export to CSV/TSV, HTML, XML
  - add -d option to dump item values and use this in test_nk2export.sh
* nk2info

Beta release
* remove deprecated definition in include/libnk2/mapi.h

20090128
* Created initial version based on libpff for the library structure and common code
* see `git log' for more recent change log



nk2tools/nk2info.c
/*
 * Shows information obtained from a Nickfile (NK2).
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "nk2tools_getopt.h"
#include "nk2tools_libcerror.h"
#include "nk2tools_libclocale.h"
#include "nk2tools_libcnotify.h"
#include "nk2tools_libnk2.h"
#include "nk2tools_output.h"
#include "nk2tools_signal.h"
#include "nk2tools_unused.h"

info_handle_t *nk2info_info_handle = NULL;
int nk2info_abort                  = 0;

/* Signal handler for nk2info
 */
void nk2info_signal_handler(
      nk2tools_signal_t signal NK2TOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "nk2info_signal_handler";

	NK2TOOLS_UNREFERENCED_PARAMETER( signal )

	nk2info_abort = 1;

	if( nk2info_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     nk2info_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use nk2info to determine information about a Nickfile (NK2).";

	nk2tools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libnk2_error_t *error      = NULL;
	system_character_t *source = NULL;
	char *program              = "nk2info";
	system_integer_t option    = 0;
	int number_of_options      = (int) ( sizeof( options ) / sizeof( nk2tools_option_t ) );
	int verbose                = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "nk2tools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( nk2tools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	nk2tools_output_version_fprint(
	 stdout,
	 program );

	if( nk2tools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = nk2tools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				nk2tools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				nk2tools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				nk2tools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		nk2tools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libnk2_notify_set_stream(
	 stderr,
	 NULL );
	libnk2_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &nk2info_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open(
	     nk2info_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     nk2info_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     nk2info_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &nk2info_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( nk2info_info_handle != NULL )
	{
		info_handle_free(
		 &nk2info_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

